src/Microsoft.Azure.CosmosRepository/Specification/Builder/OrderedSpecificationBuilder.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Copyright (c) IEvangelist. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq.Expressions;
+using System.Text;
+
+namespace Microsoft.Azure.CosmosRepository.Specification
+{
+    /// <inheritdoc cref="IOrderedSpecificationBuilder{T, TResult}"/>
+    public class OrderedSpecificationBuilder<T, TResult> : IOrderedSpecificationBuilder<T, TResult>
+        where T : IItem
+        where TResult : IQueryResult<T>
+    {
+        /// <inheritdoc/>
+        public BaseSpecification<T,TResult> Specification { get; }
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="specification"></param>
+        public OrderedSpecificationBuilder(BaseSpecification<T, TResult> specification)
+        {
+            Specification = specification;
+        }
+    }
+    /// <summary>
+    /// 
+    /// </summary>
+    public static class OrderedBuilderExtensions
+    {
+
+        /// <summary>
+        /// Adds a ThenBy expression to the existing order expression
+        /// All ThenBy expressions requires a composite index that matched the ordering clause
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <typeparam name="TResult"></typeparam>
+        /// <param name="orderBuilder"></param>
+        /// <param name="orderExpression"></param>
+        /// <returns></returns>
+        public static IOrderedSpecificationBuilder<T, TResult> ThenBy<T, TResult>(
+            this IOrderedSpecificationBuilder<T, TResult> orderBuilder,
+            Expression<Func<T, object>> orderExpression)
+            where T: IItem
+            where TResult: IQueryResult<T>
+        {
+            ((List<OrderExpressionInfo<T>>)orderBuilder.Specification.OrderExpressions)
```

Comment:
I'm not sure I like the explicit casting here, if we know the type is for sure this - why do we not expose it as such?